keps/sig-auth/20190607-certificates-api.md

```diff
@@ -132,6 +134,85 @@ A typical successful issuance proceeds as follows.
+    3. Permitted x509 extensions - none
+    4. Permitted key usages - exactly `[]string{"key encipherment", "digital signature", "client auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.  Sanity of the time is the concern of the signer.
+    6. CA bit allowed/disallowed - not allowed.
+ 3. kubernetes.io/kubelet-serving - signs serving certificates that are honored as a valid kubelet serving certificate 
+    by the kube-apiserver, but has no other guarantees.  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: signed certificates must be honored by the kube-apiserver as valid to terminate connections to a kubelet.
+       The CA bundle is not distributed by any other means.
+    2. Permitted subjects - organizations are exactly `[]string{"system:nodes"}`, common name starts with `"system:node:"`
+    3. Permitted x509 extensions - SAN DNS has value
+    4. Permitted key usages - exactly `[]string{"key encipherment", "digital signature", "server auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.
+    6. CA bit allowed/disallowed - not allowed.
+ 4. kubernetes.io/legacy-unknown - has no guarantees for trust at all.  Some distributions may honor these as client
+    certs, but that behavior is not standard kubernetes behavior.  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: None.  There is no standard trust or distribution for this signer in a kubernetes cluster.
+    2. Permitted subjects - any
+    3. Permitted x509 extensions - honors SAN extensions and discards other extensions.
+    4. Permitted key usages - may include `[]string{"key encipherment", "digital signature", "client auth", "server auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.  Sanity of the time is the concern of the signer.
+    6. CA bit allowed/disallowed - not allowed.
+
+Distribution of trust happens out of band for these signers.  Any trust outside of those described above are strictly
+coincidental.  For instance, some distributions may honor kubernetes.io/legacy-unknown as client-certificates for the
+kube-apiserver, but this is not a standard.
+None of these usages are related to ServiceAccount token secrets `.data[ca.crt]` in any way.  That ca-bundle is only
+guaranteed to verify a connection the kube-apiserver using the default service.
+
+To support HA upgrades, the kube-controller-manager will duplicate defaulting code for an empty `signerName` for one
+release.
+
+#### Limiting approval powers for certain signers.
+Given multiple signers which may be implemented as "dumb" controllers that sign if the CSR is approved, there is benefit
+to providing a simple way to subdivide approval powers through the API.  We can introduce an admission plugin that requires
+ 1. verb == `approve`
+ 2. resource == `certificatesigners`
+ 3. name == `<name of certificatesigningrequests/(.spec.signerName)>` 
```

Comment:
With `<name of certificatesigningrequests/(.spec.signerName)>` as a format, can we add a RBAC rule for e.g. `*/kubernetes.io/kubelet-client` to grant a user permission to grant *all* CertificateSigningRequests with the `kubernetes.io/kubelet-client` signerName?

Reply:
Looking here: https://github.com/kubernetes/kubernetes/blob/6099a739d8cb1762cdadf956fae484833b937c5d/pkg/apis/rbac/v1/evaluation_helpers.go#L85-L97, and with some cursory testing locally, it appears that it isn't possible to use wildcards in the `name` portion of a rule.

I think simply using the `name` as the `signerName` is probably sufficient. I don't *think* there's a use-case for per-CSR approval permissions, and right now RBAC isn't powerful enough to evaluate something like this afaict.